virttest/utils_net.py

```diff
@@ -2379,6 +2381,55 @@ def cleanup(self):
         self.close_session()
 
 
+def ieee_eui_generator(base, mask, start=0, repeat=False):
+    """
+    IEEE extended unique identifier(EUI) generator.
+
+    :param base: The base identifier number.
+    :param mask: The mask to calculate identifiers.
+    :param start: The ordinal number of the first identifier.
+    :param repeat: Whether use repeated identifiers when exhausted.
+
+    :return generator: The target EUI generator.
+    """
+    offset = 0
+    while True:
+        out = base + ((start + offset) & mask)
+        yield out
+        offset += 1
+        if offset > mask:
+            if not repeat:
+                break
+            offset = 0
+
+
+def ieee_eui48_assignment(oui_bits, prefix=0, repeat=False):
+    """
+    IEEE EUI-48 assignment.
+
+    :param oui_bits: The number of OUI bits.
+    :param prefix: The prefix of OUI, for example 0x9a.
+    :param repeat: Whether use repeated identifiers when exhausted.
+    """
+    # Using UUID1 combine with `__file__` to avoid getting the same hash
+    data = uuid.uuid1().hex + __file__
+    data = hashlib.sha256(data.encode()).digest()[:6]
+    sample = 0
+    for num in bytearray(data):
+        sample <<= 8
+        sample |= num
+    bits = 48 - oui_bits
+    mask = (1 << bits) - 1
+    start = sample & mask
+    base = sample ^ start
+    if prefix > 0:
+        pbits = 48 + (-(prefix.bit_length()) // 4) * 4
```

Comment:
doesn't this work ?
```python
pbits = 48 - (prefix.bit_length() // 4) * 4
```

Reply:
I'm afraid not, here is an example:
```python
>>> (0x7f.bit_length() // 4) * 4
4
>>> -(-0x7f.bit_length() // 4) * 4
8
```
and that because `//` stands for `floordiv`, and in this case we want `ceildiv`. It's a trick I learnt by https://stackoverflow.com/questions/2356501/how-do-you-round-up-a-number-in-python#answer-35125872

Reply:
cool, Thanks for explaining.